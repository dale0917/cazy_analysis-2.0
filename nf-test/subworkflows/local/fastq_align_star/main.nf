include { STAR_GENOMEGENERATE    } from '/home/zs-dalo/modules/modules/modules/zs/star/genomegenerate/main.nf'
include { RSEM_PREPAREREFERENCE  } from '/home/zs-dalo/modules/modules/modules/zs/prepare_reference/preparereference.nf'
include { STAR_ALIGN             } from '/home/zs-dalo/modules/modules/modules/zs/star/align/main.nf'

workflow FASTQ_ALIGN_STAR {

    take:
    ch_reads
    ch_index  // if not using, should pass a '[]' so that subworkflow can run and perform conditionals
    ch_fasta  // if not using, should pass a '[]' so that subworkflow can run and perform conditionals
    ch_gtf    // if not using, should pass a '[]' so that subworkflow can run and perform conditionals

    main:
    ch_versions    = Channel.empty()
    ch_final_index = Channel.empty()

    // Si no se proporciona un índice pre-construido
    
    if (!ch_index) {
        assert (ch_fasta && ch_gtf), "If not using a pre-built index, a fasta and gtf reference files are required for running the FASTQ_ALIGN_STAR subworkflow"

        // Genera el índice del genoma con STAR
        STAR_GENOMEGENERATE(
            ch_fasta,
            ch_gtf
        )
    }
    ch_versions = ch_versions.mix(STAR_GENOMEGENERATE.out.versions)

    if (!ch_index) {
        assert (ch_fasta && ch_gtf), "If not using a pre-built index, a fasta and gtf reference files are required for running the FASTQ_ALIGN_STAR subworkflow"

        RSEM_PREPAREREFERENCE (ch_fasta, ch_gtf)
        ch_final_index = RSEM_PREPAREREFERENCE.out.index

    } else {
        ch_final_index = Channel.value(file(index))
    }

    ch_versions = ch_versions.mix(RSEM_PREPAREREFERENCE.out.versions)

    STAR_ALIGN ( 
        ch_reads,
        ch_final_index
    )

    //ch_versions = ch_versions.mix(STAR_ALIGN.out.versions)

    emit:
    index     = ch_final_index
    bam       = STAR_ALIGN.out.bam
    log_final = STAR_ALIGN.out.log_final
    versions  = ch_versions
    

}

