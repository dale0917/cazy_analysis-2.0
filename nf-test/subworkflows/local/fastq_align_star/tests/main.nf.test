nextflow_workflow {

    name "Test Subworkflow FASTQ_ALIGN_STAR"
    script "../main.nf"
    config "./nextflow.config"
    workflow "FASTQ_ALIGN_STAR"

    tag "subworkflows"
    tag "subworkflows/fastq_align_star"
    tag "fastq_align_star"

    setup {
        run("STAR_GENOMEGENERATE") {
            script "../../../../modules/zs/star/genomegenerate/main.nf"
            process {
                """
                input[0] = Channel.of([
                    [ id:'test_fasta' ],
                    [ file(params.test_data['homo_sapiens']['genome']['genome_fasta'], checkIfExists: true) ]
                ])
                input[1] = Channel.of([
                    [ id:'test_gtf' ],
                    [ file(params.test_data['homo_sapiens']['genome']['genome_gtf'], checkIfExists: true) ]
                ])
                """
            }
        }
    }

    test("star with pre-built index - paired_end") {
        // using nf-test setup() method to pre-built and index for workflow

        when {
            workflow { 
                """
                input[0] = Channel.of([
                    [ id:'nfcore-test-sample', single_end:false ],
                    [
                        file(params.test_data['homo_sapiens']['illumina']['test_1_fastq_gz'], checkIfExists: true),
                        file(params.test_data['homo_sapiens']['illumina']['test_2_fastq_gz'], checkIfExists: true)
                    ]
                ])
                input[1] = STAR_GENOMEGENERATE.out.index.map{ it[1] }
                input[2] = []
                input[3] = []
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out.bam).match() } // only file that stays the same
            )
        }

    }

    test("star without pre-built index - paired_end") {

        when {
            workflow { 
                """
                input[0] = Channel.of([
                    [ id:'nfcore-test-sample', single_end:false ],
                    [
                        file(params.test_data['homo_sapiens']['illumina']['test_1_fastq_gz'], checkIfExists: true),
                        file(params.test_data['homo_sapiens']['illumina']['test_2_fastq_gz'], checkIfExists: true)
                    ]
                ])
                input[1] = []
                input[2] = Channel.of([
                    [ id:'test_fasta' ],
                    [ file(params.test_data['homo_sapiens']['genome']['genome_fasta'], checkIfExists: true) ]
                ])
                input[3] = Channel.of([
                    [ id:'test_gtf' ],
                    [ file(params.test_data['homo_sapiens']['genome']['genome_gtf'], checkIfExists: true) ]
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out.bam).match() } // only file that stays the same
            )
        }

    }

}

