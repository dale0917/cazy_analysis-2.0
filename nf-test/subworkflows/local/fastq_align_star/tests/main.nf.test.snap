{
    "star without pre-built index - paired_end": {
        "content": [
            [
                [
                    {
                        "id": "nfcore-test-sample",
                        "single_end": false
                    },
                    "nfcore-test-sample.Aligned.toTranscriptome.out.bam:md5,f32289e2a23390542b4dd40da941193d"
                ]
            ]
        ],
        "timestamp": "2024-03-26T13:07:14.73782"
    },
    "star with pre-built index - paired_end": {
        "content": [
            [
                [
                    {
                        "id": "nfcore-test-sample",
                        "single_end": false
                    },
                    "nfcore-test-sample.Aligned.toTranscriptome.out.bam:md5,f32289e2a23390542b4dd40da941193d"
                ]
            ]
        ],
        "timestamp": "2024-03-26T13:06:39.036651"
    }
}